pkg/kubectl/cmd/logs.go

```diff
@@ -196,26 +205,73 @@ func (o *LogsOptions) Complete(f cmdutil.Factory, out io.Writer, cmd *cobra.Comm
 		}
 	}
 
-	mapper, typer := f.Object()
-	decoder := f.Decoder(true)
+	o.Mapper, o.Typer = f.Object()
+	o.Decoder = f.Decoder(true)
 	if o.Object == nil {
-		builder := resource.NewBuilder(mapper, f.CategoryExpander(), typer, o.ClientMapper, decoder).
+		builder := resource.NewBuilder(o.Mapper, f.CategoryExpander(), o.Typer, o.ClientMapper, o.Decoder).
 			NamespaceParam(o.Namespace).DefaultNamespace().
 			SingleResourceType()
 		if o.ResourceArg != "" {
 			builder.ResourceNames("pods", o.ResourceArg)
 		}
-		if selector != "" {
-			builder.ResourceTypes("pods").SelectorParam(selector)
+		if o.Selector != "" {
+			builder.ResourceTypes("pods").SelectorParam(o.Selector)
 		}
 		infos, err := builder.Do().Infos()
+		// if name not found use prefix match
+		aggErrs := []error{}
 		if err != nil {
-			return err
+			aggErrs = append(aggErrs, err)
+			if apierrors.IsNotFound(err) && o.ResourceArg != "" {
+				err = o.prefixMatchResourceList(simpleTrimResourceTypeName(o.ResourceArg), err)
+				if err != nil {
+					aggErrs = append(aggErrs, err)
+					return utilerrors.NewAggregate(aggErrs)
+				}
+				if len(o.PrefixMatchList) > 0 {
+					return nil
+				}
+			}
 		}
-		if selector == "" && len(infos) != 1 {
+		if o.Selector == "" && len(infos) != 1 {
 			return errors.New("expected a resource")
 		}
-		o.Object = infos[0].Object
+
+		if o.Object == nil {
+			o.Object = infos[0].Object
+		}
+	}
+	return nil
+}
+
+func simpleTrimResourceTypeName(s string) string {
+	if strings.Contains(s, "/") {
+		seg := strings.Split(s, "/")
+		return seg[0]
+	}
+	return "pods"
+}
+
+func (o *LogsOptions) prefixMatchResourceList(resourceType string, originalError error) error {
+	infos, err := resource.NewBuilder(o.Mapper, o.CategoryExpander, o.Typer, o.ClientMapper, o.Decoder).
+		NamespaceParam(o.Namespace).DefaultNamespace().
+		SingleResourceType().ResourceTypeOrNameArgs(true, resourceType).
+		Flatten().Do().Infos()
+	if err != nil {
+		return err
+	}
+	//match pods using prefix
+	prefixTimes := 0
+	for i, info := range infos {
+		if strings.HasPrefix(info.Name, o.ResourceArg) {
+			o.PrefixMatchList = append(o.PrefixMatchList, info.Name)
```

Comment:
Mutating the parent object like this is not part of our normal style.  This should be a return value from this function, rather than done here.  It would be better if this method perform the entire behavior of your alternate lookup.